src/components/implementation/no_interface/scheddev/sl.c

```diff
@@ -61,40 +83,103 @@ sl_thd_block(thdid_t tid)
 
 	sl_cs_enter();
 	t = sl_thd_curr();
-	if (unlikely(t->state == SL_THD_WOKEN)) {
-		t->state = SL_THD_RUNNABLE;
+	if (sl_thd_block_no_cs(t, 0)) {
 		sl_cs_exit();
 		return;
 	}
-
-	assert(t->state == SL_THD_RUNNABLE);
-	t->state = SL_THD_BLOCKED;
-	sl_mod_block(sl_mod_thd_policy_get(t));
 	sl_cs_exit_schedule();
 
 	return;
 }
 
-void
-sl_thd_wakeup(thdid_t tid)
+cycles_t
+sl_thd_block_timeout(thdid_t tid, cycles_t abs_timeout)
 {
+	cycles_t jitter = 0;
 	struct sl_thd *t;
-	thdcap_t       thdcap;
-	tcap_t         tcap;
-	tcap_prio_t    prio;
+	int is_timeout = abs_timeout ? 1 : 0;
+
+	/* TODO: dependencies not yet supported */
+	assert(!tid);
 
 	sl_cs_enter();
-	t = sl_thd_lkup(tid);
-	if (unlikely(!t)) goto done;
+	t = sl_thd_curr();
+	if (sl_thd_block_no_cs(t, is_timeout)) {
+		sl_cs_exit();
+		return 0;
+	}
+	if (is_timeout) sl_timeout_block(t, abs_timeout);
+	sl_cs_exit_schedule();
+
+	/* FIXME: possible race around these accesses */
+	if (t->wakeup_cycs > t->timeout_cycs) jitter = t->wakeup_cycs - t->timeout_cycs;
```

Comment:
OK, I was worried about incorrectness here.  The only data-structure we're accessing is the current thread, and if we're executing here, that thread structure should be valid.  Additionally, since we're only *accessing* memory (not changing it), it seems we're likely to avoid races.

Should be OK, but the analysis to determine that should be taken and is pretty nuanced.